File: _6_Coordinate_EP/dIntraInter/boss.c
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#define MAX_MSG_LEN 128
void spawn_children(void);

int main(int argc, char** argv) {

    MPI_Init(&argc, &argv);

    spawn_children();
    MPI_Finalize();
    return 0;
}

void spawn_children(void) {
    int rank, size;
    char* command = "./child.exe";
    char msgstr[MAX_MSG_LEN];
    MPI_Status status;
    MPI_Comm child_comm;

    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);


    MPI_Comm_spawn(command, MPI_ARGV_NULL, 3, MPI_INFO_NULL, 0, MPI_COMM_WORLD, &child_comm, MPI_ERRCODES_IGNORE);
    MPI_Send("Hello", MAX_MSG_LEN, MPI_CHAR, 2, 0, child_comm);
    printf("Parent %d sent %s to %d of comm %d\n", rank, "Hello", 2, child_comm);
    if (rank == 1) {
        MPI_Recv(&msgstr, MAX_MSG_LEN, MPI_CHAR, MPI_ANY_SOURCE, MPI_ANY_TAG, child_comm, &status);
        printf("Parent %d received %s from %d of comm %d\n", rank, msgstr, status.MPI_SOURCE, child_comm);
//        Intra from rank 0 to 1
        MPI_Recv(&msgstr, MAX_MSG_LEN, MPI_CHAR, MPI_ANY_SOURCE, MPI_ANY_TAG, MPI_COMM_WORLD, &status);
        printf("Parent Intra %d received %s from %d of comm %d\n", rank, msgstr, status.MPI_SOURCE, MPI_COMM_WORLD);
    } else if (rank == 0) {
        MPI_Send("Rank 0 to 1", MAX_MSG_LEN, MPI_CHAR, 1, 0, MPI_COMM_WORLD);
    }
}

File: _6_Coordinate_EP/gSync/boss.f90
program mpi_app
    use mpi
    implicit none

    call spawn_children()

contains

    subroutine spawn_children()
        implicit none
        include 'mpif.h'
        integer :: ierr, rank, num_procs, new_comm, parent_comm, child_idx,status(MPI_STATUS_SIZE)
        integer ::  calling = 0, num_children = 1
        real(kind=8) :: received_data(2), random_oat_c
        logical :: inited
        character(len=50) :: command
        !define one global variable countCalling
        print *, "calling", calling
        if (calling > 0) then
            return
        end if
        calling = calling + 1

        call MPI_Initialized(inited, ierr)
        if (.not. inited) then
            call MPI_Init(ierr)
        end if
        call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)
        call MPI_Comm_size(MPI_COMM_WORLD, num_procs, ierr)
        print *, "MPI_COMM_WORLD", MPI_COMM_WORLD, "num_procs", num_procs, "rank", rank

        call MPI_Comm_dup(MPI_COMM_WORLD, parent_comm, ierr)
        command = "./child.exe"
        call MPI_Comm_spawn(command, MPI_ARGV_NULL, num_children, MPI_INFO_NULL, &
                0, parent_comm, new_comm, MPI_ERRCODES_IGNORE, ierr)
        print *, "Parent comm: ", parent_comm, " new comm: ", new_comm

        do while (.true.)
            do child_idx = 1, num_children
                call random_number(random_oat_c)
                random_oat_c = 12 + int(random_oat_c * 28)
                call MPI_Send(random_oat_c, 1, MPI_REAL8, child_idx-1, 0, new_comm, ierr)
                call MPI_Recv(received_data(child_idx), 1, MPI_REAL8, child_idx-1, MPI_ANY_TAG, new_comm, status, ierr)
            end do
            print *, "received_data", received_data
            if (status(MPI_TAG) == 886) then
                print *, "ending messsage 886 received, exiting the loop"
                exit
            end if
        end do

        call MPI_Barrier(new_comm, ierr)
        call MPI_Comm_free(new_comm, ierr)
        call MPI_Comm_free(parent_comm, ierr)
        call MPI_Finalize(ierr)

    end subroutine spawn_children

end program mpi_app
